hudi-common/src/test/java/org/apache/hudi/common/functional/TestHoodieLogFormat.java

```diff
@@ -671,11 +658,188 @@ public void testBasicAppendAndScanMultipleFiles(ExternalSpillableMap.DiskMapType
+    List<IndexedRecord> scannedAvroRecords = new ArrayList<>();
+    for (HoodieRecord record : scanner) {
+      scannedHoodieRecords.add(record);
+      scannedAvroRecords.add((IndexedRecord)
+          ((HoodieAvroRecord) record).getData().getInsertValue(schema).get());
+    }
+
+    assertEquals(sort(sampledRecords), sort(scannedAvroRecords));
+
+    //
+    // Step 2: Scan by the same list of keys (no new scanning should be performed,
+    //         in this case, and same _objects_ have to be returned)
+    //
+
+    scanner.scanByFullKeys(sampledRecordKeys);
+
+    List<HoodieRecord> newScannedHoodieRecords = new ArrayList<>();
+    for (HoodieRecord record : scanner) {
+      newScannedHoodieRecords.add(record);
+    }
+
+    assertEquals(scannedHoodieRecords.size(), newScannedHoodieRecords.size());
+
+    for (int i = 0; i < scannedHoodieRecords.size(); ++i) {
+      assertSame(scannedHoodieRecords.get(i), newScannedHoodieRecords.get(i), "Objects have to be identical");
+    }
+
+    scanner.close();
+  }
+
+  @ParameterizedTest
+  @MethodSource("testArguments")
+  public void testBasicAppendAndPartialScanningByKeyPrefixes(ExternalSpillableMap.DiskMapType diskMapType,
+                                                             boolean isCompressionEnabled,
+                                                             boolean readBlocksLazily,
+                                                             boolean useScanV2)
+      throws IOException, URISyntaxException, InterruptedException {
+    // Generate 3 delta-log files w/ random records
+    Schema schema = HoodieAvroUtils.addMetadataFields(getSimpleSchema());
+    List<IndexedRecord> genRecords = SchemaTestUtil.generateHoodieTestRecords(0, 300);
+
+    Set<HoodieLogFile> logFiles = writeLogFiles(partitionPath, schema, genRecords, 3);
+
+    FileCreateUtils.createDeltaCommit(basePath, "100", fs);
+
+    // scan all log blocks (across multiple log files)
+    HoodieMergedLogRecordScanner scanner = HoodieMergedLogRecordScanner.newBuilder()
+        .withFileSystem(fs)
+        .withBasePath(basePath)
+        .withLogFilePaths(
+            logFiles.stream()
+                .map(logFile -> logFile.getPath().toString()).collect(Collectors.toList()))
+        .withReaderSchema(schema)
+        .withLatestInstantTime("100")
+        .withMaxMemorySizeInBytes(1024L)
+        .withReadBlocksLazily(readBlocksLazily)
+        .withReverseReader(false)
+        .withBufferSize(BUFFER_SIZE)
+        .withSpillableMapBasePath(spillableBasePath)
+        .withDiskMapType(diskMapType)
+        .withBitCaskDiskMapCompressionEnabled(isCompressionEnabled)
+        .withUseScanV2(useScanV2)
+        .withRecordMerger(HoodieRecordUtils.loadRecordMerger(HoodieAvroRecordMerger.class.getName()))
+        .withForceFullScan(false)
+        .build();
+
+
+    List<String> sampledRecordKeys = Arrays.asList(
+        "00509b14-3d1a-4283-9a8c-c72b971a9d06",
```

Comment:
is it not possible to generate the list dynamically. would prefer to avoid hard coded keys/records. may be while generating the records, is it possible to dictate some prefix.

Reply:
Discussed offline: we actually want to keep tests reproducible, which will help in debugging them